packages/web-components/src/avatar/avatar.ts

```diff
@@ -0,0 +1,165 @@
+  AvatarColor,
+  AvatarNamedColor,
+  AvatarShape,
+  AvatarSize,
+} from './avatar.options.js';
+
+/**
+ * The base class used for constructing a fluent-avatar custom element
+ * @public
+ */
+export class Avatar extends FASTElement {
+  /**
+   * The name of the person or entity represented by this Avatar. This should always be provided if it is available.
+   *
+   * @public
+   * @remarks
+   * HTML Attribute: name
+   */
+  @attr
+  public name: string | undefined;
+
+  /**
+   * Provide custom initials rather than one generated via the name
+   *
+   * @public
+   * @remarks
+   * HTML Attribute: name
+   */
+  @attr
+  public initials: string | undefined;
+
+  /**
+   * Size of the avatar in pixels.
+   *
+   * Size is restricted to a limited set of supported values recommended for most uses (see `AvatarSizeValue`) and
+   * based on design guidelines for the Avatar control.
+   *
+   * If a non-supported size is neeeded, set `size` to the next-smaller supported size, and set `width` and `height`
+   * to override the rendered size.
+   *
+   * @public
+   * @remarks
+   * HTML Attribute: size
+   *
+   */
+  @attr({ converter: nullableNumberConverter })
+  public size: AvatarSize | undefined;
+
+  /**
+   * The avatar can have a circular or square shape.
+   *
+   * @public
+   * @remarks
+   * HTML Attribute: shape
+   */
+  @attr
+  public shape: AvatarShape | undefined;
+
+  /**
+   * Optional activity indicator
+   * * active: the avatar will be decorated according to activeAppearance
+   * * inactive: the avatar will be reduced in size and partially transparent
+   * * undefined: normal display
+   *
+   * @public
+   * @remarks
+   * HTML Attribute: active
+   */
+  @attr
+  public active?: AvatarActive | undefined;
+
+  /**
+   * The appearance when `active="active"`
+   *
+   * @public
+   * @remarks
+   * HTML Attribute: appearance
+   */
+  @attr
+  public appearance?: AvatarAppearance | undefined;
+
+  /**
+   * The color when displaying either an icon or initials.
+   * * neutral (default): gray
+   * * brand: color from the brand palette
+   * * colorful: picks a color from a set of pre-defined colors, based on a hash of the name (or colorId if provided)
+   * * [AvatarNamedColor]: a specific color from the theme
+   *
+   * @public
+   * @remarks
+   * HTML Attribute: color
+   */
+  @attr
+  public color: AvatarColor = 'neutral';
+
+  /**
+   * Specify a string to be used instead of the name, to determine which color to use when color="colorful".
+   * Use this when a name is not available, but there is another unique identifier that can be used instead.
+   */
+  @attr({ attribute: 'color-id' })
+  public colorId: AvatarNamedColor | undefined;
+  public colorIdChanged(): void {
+    if (!this.colorId) {
+      return;
+    }
+
+    this.generateColor();
+  }
+
+  /**
+   * Sets the data-color attribute used for the visual presentation
+   * @internal
+   */
+  public generateColor(): AvatarColor | void {
```

Comment:
> > I think all the *Changed callbacks which call either setInitials() or setColor() are unnecessary asboth are just compute functions which do not alter anything. When any of name, initials, color or color-id attributes changes, the relevant part of the template is invalidated and re-rendered.
>
> Noted elsewhere but the one which doesn't change is color-id, there's nothing that triggers a change unless you are changing the value of color, if it's already colorful it's there's nothing to change, so that one remains.

I am sorry, but I still do not follow. I would like to understand to learn.

The `generateColor()` just generates the `AvatarColor` without any side-effects (besides calling property getters).
It is called from the template, which, according to the [Fast Element docs](https://www.fast.design/docs/fast-element/observables-and-state/#access-tracking) tracks access to the `@attr`s. That mean's that once the relevant attribute changes, (part of) the template is re-rendered.

The `colorId` getter in this function is called inside a condition, but that should not be an issue?

Now, let's imagine that the template fails to track the changes to `colorId`. How is calling the `generateColor()` from `colorIdChanged()` supposed to help?

Reply:
This is what I see with the `colorIdChanged()` commented out. I start with `color: neutral` (default), I set `color: colorful` -> template re-renders, I set `color-id: beige` -> template re-renders again, `data-color` is set.

<img width="935" alt="image" src="https://user-images.githubusercontent.com/9615899/220577933-d533c734-4a04-424c-8bf0-ac7fb11fe4f3.png">

Everything still seems to work as expected. What is the use case which breaks?

Reply:
No, you're right - must've been the middle of my long day :). Removed and resolved.